### PR TITLE
New context attributes syntax

### PR DESCRIPTION
**Pull request by petergebala**
Co-authored-by: @kikorb 

Added new syntax for `context|params|dependencies`
Part of: https://github.com/Profinda/opera/issues/12
```ruby
context do
  attr_accessor :custom_value
end

dependencies do
  attr_reader :account, :requester
end

alias_method :requester, :profile
```

We still keep backward compatibility till next major version

### CODE REVIEW

**Review by kikorb — APPROVED** [review]

**Comment by cintrzyk** [comment]
> We still keep backward compatibility till next major version

does it mean that helpers like `context_reader`, `context_accessor`, `dependencies_reader`, etc. going to be deprecated?

I'm not entirely sure how the proposed syntax improves upon our current one 🤔 Could you explain the reasoning behind it in more detail?

**Comment by petergebala** [comment]
For now we keep backward compatibility. Plan is to get rid of it in next version.
New syntax will work better with modern IDE and autocompletion.

**Comment by astoica1986** [comment]
Wdyt about adding a little syntactic sugar and supporting this for schema outputs as well? :) 
`schema do`
`  attr_reader :user_id`
`end`

**Review by astoica1986 — APPROVED** [review]

**Comment by petergebala** [comment]
> Wdyt about adding a little syntactic sugar and supporting this for schema outputs as well? :) 
> `schema do`
> `  attr_reader :user_id`
> `end`

Might be hard. Schema is our concept. Name of validate step. Is dynamic and you may have multiple validations inside single operation

**Comment by cintrzyk** [comment]
I see your point about compatibility with IDEs, but I believe there’s more to consider 🤔 
Currently, using helpers like `context_reader :x` clearly indicates that we’re defining a getter specific to the `context` object, which makes it immediately understandable. Using the same method names as ruby, but having them do something different, just adds unnecessary complexity (for the sake of "tricking" the IDE?). If I understand correctly, the new syntax would work better with the IDE precisely because we're reusing the same naming as in ruby attribute accessors?

**Comment by kikorb** [comment]
You are used to the old one. Be people coding will immediately understand
attr_reader and doing so inside context or dependencies block will make
them connect the dots.

I thought this would be better and make things clear but let's discuss it
on the next roadmap call

**Comment by petergebala** [comment]
> I see your point about compatibility with IDEs, but I believe there’s more to consider 🤔 
> Currently, using helpers like `context_reader :x` clearly indicates that we’re defining a getter specific to the `context` object, which makes it immediately understandable. Using the same method names as ruby, but having them do something different, just adds unnecessary complexity (for the sake of "tricking" the IDE?). If I understand correctly, the new syntax would work better with the IDE precisely because we're reusing the same naming as in ruby attribute accessors? 

Name is the same on purpose to get for free LSP support. You can still define ruby acessors if needed (not preferable in operation) outside of block.
What is more we worte it in such a way that we don't modify Ruby method(class level) . We define them as instance in isolated new class

**Comment by astoica1986** [comment]
To @cintrzyk's point, as a new user if I see something like:
`attr_reader :account` I might immediately assume there is an @account instance var defined somewhere. 
I don't see that as much of a problem but I've used Opera for years now and it all seems completely natural to me.

**Comment by cintrzyk** [comment]
using syntax like `attr_reader :example, default: -> { 'xxxx' }` is confusing to me because `attr_reader` in ruby doesn't accept options 😅 let's chat on the retro 👍

File: spec/spec_helper.rb [hunk 1]
```diff
@@ -1,5 +1,6 @@
 require "bundler/setup"
 require "opera"
+require "pry"
```
Comment (javiyu) [inline spec/spec_helper.rb:3]: Was this intended? I'm OK having it if it was intended but the majority of specs runs are automated and would not use pry. Not sure how much overhead it adds to the pipeline though.

File: .github/workflows/release.yml [hunk 2]
```diff
@@ -9,7 +9,7 @@ jobs:
     runs-on: ubuntu-latest
     strategy:
       matrix:
-        ruby-version: [ '2.6', '2.7', '3.0', '3.2', 'jruby' ]
```
Comment (javiyu) [inline .github/workflows/release.yml:12]: I would approve both changes, but we are mixing two different things on this PR:
- dependency upgrade
- new syntax

**Review by javiyu — APPROVED** [review]

**Comment by petergebala** [comment]
I will not be able to attend meeting 🙁 my 2 cents:
1. Ruby's attr accessor defines getter and setter to access internals. We do the same.
2. Extra defualt setting is self explanatory. Does not need extra explanation and is known in Ruby community. What is more we have documentation with examples.

**Pull request merged**